woocommerce/woocommerce-gutenberg-products-block
# Create typedefs and add basic typescript configuration

**NOTE:** This pull will NOT require us starting to write our code in typescript. It simply is adding typescript support that will enhance our current usage of jsdoc.

Lately I've been playing with the typescript configuration in the Gutenberg project and have discovered it [greatly helps with improving the quality of code](https://andrewduthie.com/2019/11/27/typescript-tooling-for-your-javascript-projects/) - including things like:

- catching syntax errors (that linting can't pick up)
- ensuring types are properly used.
- better code completion in IDE's (vscode has GREAT typescript support).
- pointing out improper property usage where we declare types (catching typos in property names etc).

With our work increasing in complexity, I want to contribute to improving our code quality so I've created this pull to help us incrementally get there.  With minimal adjustments this shouldn't affect our current velocity but does help us to start opting in over time some improvements to our syntax to enable better code quality.

In this pull, I've introduced:

- installing typescript, and adding a basic configuration that should handle our aliased module imports.
- started a `type-defs` folder with an initial set of `cart` typedefs that I'm using in the payment methods work I'm doing. These cart typedefs will be usable elsewhere too and I'd like us to start using them where we can (instructions coming below). We can start using this folder for holding all typedefs.
- Completed a minor fix in `assets/js/blocks/cart-checkout/cart/full-cart/index.js` that was exposed by the typescript config! This wouldn't have affected anything but still is a fix nevertheless.

## Setting things up and usage.

If you're using VScode, you shouldn't have to do anything!  

**Usage**

If you are using standard `jsdoc` notation, the typescript configuration will automatically derive hints from the jsdoc implementation. However, I'd like us to start using declared type definitions for commonly used objects (like `CartShippingItem` etc).  That way instead of doing things like this in our jsdoc implementation:

```jsx
/**
 * @param {Object} cartAddress 
 */
const somethingUsingCart = ( cartAddress ) => {
  const firstName = cartAddress.first_name;
}
```

We can instead do this:

```jsx
/** typedef {import('@woocommerce/type-defs/cart').CartShippingAddress} CartShippingAddress */

/**
 * 
 * @param {CartShippingAddress} cartAddress 
 */
const somethingUsingCart = ( cartAddress ) => {
  const firstName = cartAddress.first_name;
}
```

This would then get us type checking and code completion automatically within our IDE.

## Next Steps

- review this
- merge
- I may start recommending usage of better jsdocs in code reviews I do, but will not be sticky about it yet.
- You will start seeing Typescript warn (red lines!) about code issues in some files as you work on them. Note the errors by hovering over them and you'll have hints on how to resolve. None of these warnings should affect building code or fail tests etc. But they would be good to heed because there's good reason for them. Ping me if something seems unresolveable and I'll assist.
- I'll write a p2 highlighting what this enables and how to get the most out of it.
- Personally, I'm going to be starting to implement more typedefs in code I write and implement (providing some examples for how to use and benefits it brings).
- Over the cooldown, I want us as a squad to start implementing more thoroughly in code we write (being more intentional about using jsdoc and typedefs for typed objects).

So ultimately, this shouldn't affect velocity of the team just yet but does lay the groundwork for adopting better code quality checking both now and in the future.

## To Test

The work done in this pull should not affect any builds or tests so those should all pass. But you can do a smoke test of both `npm run start` and `npm run build` to ensure both dev and production builds built properly and there's nothing obviously broken in existing functionality.
